hail/python/hailtop/aiotools/s3asyncfs.py

```diff
@@ -0,0 +1,413 @@
+                                                 **self._kwargs)
+            return self._page
+
+        next_continuation_token = self._page.get('NextContinuationToken')
+        if next_continuation_token is not None:
+            self._page = await blocking_to_async(self._fs._thread_pool, self._fs._s3.list_objects_v2,
+                                                 Bucket=self._bucket,
+                                                 Prefix=self._prefix,
+                                                 ContinuationToken=next_continuation_token,
+                                                 **self._kwargs)
+            return self._page
+
+        raise StopAsyncIteration
+
+
+class S3HeadObjectFileStatus(FileStatus):
+    def __init__(self, head_object_resp):
+        self.head_object_resp = head_object_resp
+
+    async def size(self) -> int:
+        return self.head_object_resp['ContentLength']
+
+    async def __getitem__(self, key: str) -> Any:
+        return self.head_object_resp[key]
+
+
+class S3ListFilesFileStatus(FileStatus):
+    def __init__(self, item: Dict[str, Any]):
+        self._item = item
+
+    async def size(self) -> int:
+        return self._item['Size']
+
+    async def __getitem__(self, key: str) -> Any:
+        return self._item.get(key)
+
+
+class S3CreateManager(AsyncContextManager[WritableStream]):
+    def __init__(self, fs, bucket, name):
+        self.fs = fs
+        self.bucket = bucket
+        self.name = name
+        self.async_writable = None
+        self.put_task = None
+        self._value = None
+
+    async def __aenter__(self) -> WritableStream:
+        async_writable, blocking_readable = async_writable_blocking_readable_stream_pair()
+        self.async_writable = async_writable
+        self.put_task = asyncio.create_task(
+            blocking_to_async(self.fs._thread_pool, self.fs._s3.upload_fileobj,
+                              blocking_readable,
+                              Bucket=self.bucket,
+                              Key=self.name))
+        return async_writable
+
+    async def __aexit__(
+            self, exc_type: Optional[Type[BaseException]] = None,
+            exc_value: Optional[BaseException] = None,
+            exc_traceback: Optional[TracebackType] = None) -> None:
+        await self.async_writable.wait_closed()
+        self._value = await self.put_task
+
+
+class S3FileListEntry(FileListEntry):
+    def __init__(self, bucket: str, key: str, item: Optional[Dict[str, Any]]):
+        assert key.endswith('/') == (item is None)
+        self._bucket = bucket
+        self._key = key
+        self._item = item
+        self._status: Optional[S3ListFilesFileStatus] = None
+
+    def name(self) -> str:
+        return os.path.basename(self._key)
+
+    async def url(self) -> str:
+        return f's3://{self._bucket}/{self._key}'
+
+    def url_maybe_trailing_slash(self) -> str:
+        return f's3://{self._bucket}/{self._key}'
```

Comment:
So I think having these two functions is bad, and I'd like to revise the code (although not in this PR).

A little background, there can be some ambiguity, esp. on key-indexed value stores, whether something like "/foo/bar" is a file or a directory (in fact on GCS it can be both).  A second issue is that paths can have empty components in GCS, for example, "/foo//bar" is a valid path.  Then "/foo/bar/" is ambiguous: does it mean the directory named "bar" located in "/foo", or does it mean the directory with an empty name "" located in "/foo/bar"?

So I decided inside the AsyncFS library to always pass directories with the trailing slash to remove this ambiguity.  However, there are situations when you need to do an additional system call or HTTP call to determine if something is a directory, and I really tried to design the library to only require the minimal number of system calls or API calls per operation.

It turned out, sometimes I needed the path, but didn't want to perform an additional operation, so I added url_maybe_trailing_slash (that is, url, with a bonus trailing slash if we happen to know whether the url is a directory or not).

In the S3 backend, we always know if something is a directory.  See the line above in the constructor:

 > `assert key.endswith('/') == (item is None)`

Therefore, they are the same because we're always including the (optional) trailing slash.

That help?

Reply:
thanks, that helps!